### PR TITLE
Bump eslint plugins

### PR DESCRIPTION
**Pull request by BPScott**
Fixed: #161

Preps for the latest major version of TS tooling, and sets the stage for
ESLint 7 updates, supporting optional-chaining

Major bumps are:

- @typescript-eslint/eslint-plugin to 4.1.0
- eslint-plugin-graphql to 4.0.0
- eslint-plugin-jest to 24.0.0

### CODE REVIEW

File: packages/eslint-plugin/lib/config/typescript-type-checking.js [hunk 1]
```diff
@@ -34,10 +34,7 @@ module.exports = {
         // When adding two variables, operands must both be of type number or of type string
         '@typescript-eslint/restrict-plus-operands': 'error',
         // Prevents conditionals where the type is always truthy or always falsy
-        '@typescript-eslint/no-unnecessary-condition': [
-          'error',
-          {ignoreRhs: true},
-        ],
+        '@typescript-eslint/no-unnecessary-condition': ['error'],
```
Comment (BPScott) [inline packages/eslint-plugin/lib/config/typescript-type-checking.js:37]: The `ignoreRhs` was removed

File: packages/eslint-plugin/lib/config/rules/jest.js [hunk 2]
```diff
@@ -44,7 +44,7 @@ module.exports = {
   // Disallow setup and teardown hooks
   'jest/no-hooks': 'off',
   // Using a callback in asynchronous tests
-  'jest/no-test-callback': 'error',
+  'jest/no-done-callback': 'error',
```
Comment (alexandcote) [inline packages/eslint-plugin/lib/config/rules/jest.js:47]: https://github.com/jest-community/eslint-plugin-jest/pull/653

File: packages/eslint-plugin/package.json [hunk 3]
```diff
@@ -33,25 +33,25 @@
...
     "change-case": "^4.1.1",
     "common-tags": "^1.8.0",
     "eslint-config-prettier": "6.11.0",
     "eslint-module-utils": "2.6.0",
-    "eslint-plugin-babel": "5.3.0",
-    "eslint-plugin-eslint-comments": "3.1.2",
-    "eslint-plugin-graphql": "3.1.1",
+    "eslint-plugin-babel": "5.3.1",
+    "eslint-plugin-eslint-comments": "3.2.0",
+    "eslint-plugin-graphql": "4.0.0",
```
Comment (alexandcote) [inline packages/eslint-plugin/package.json:45]: Just double-checking, this version only supports Node.js 10.

File: packages/eslint-plugin/lib/config/rules/jest.js (hunk 2 above)
Comment (alexandcote) [reply packages/eslint-plugin/lib/config/rules/jest.js:47]: `jest/no-try-expect` is deprecated in favor of `jest/no-conditional-expect`

Comment (alexandcote) [reply packages/eslint-plugin/lib/config/rules/jest.js:47]: Should we add those new rules
- `jest/no-interpolation-in-snapshots`
- `jest/no-deprecated-functions`

File: yarn.lock [hunk 4]
```diff
@@ -890,15 +897,22 @@
...
 
-"@babel/runtime@^7.4.5", "@babel/runtime@^7.8.4", "@babel/runtime@^7.8.7":
+"@babel/runtime@^7.10.0", "@babel/runtime@^7.10.2":
+  version "7.11.2"
+  resolved "https://registry.yarnpkg.com/@babel/runtime/-/runtime-7.11.2.tgz#f549c13c754cc40b87644b9fa9f09a6a95fe0736"
+  integrity sha512-TeWkU52so0mPtDcaCTxNBI/IHiz0pZgr8VEFqXFtZWpYD08ZB6FaSwVAS8MKRQAP3bYKiVjwysOJgMFY28o6Tw==
+  dependencies:
+    regenerator-runtime "^0.13.4"
+
+"@babel/runtime@^7.8.4", "@babel/runtime@^7.8.7":
```
Comment (alexandcote) [inline yarn.lock:915]: Dedupe

File: yarn.lock [hunk 5]
```diff
@@ -4195,6 +4427,18 @@ globby@*, globby@^11.0.0:
     merge2 "^1.3.0"
     slash "^3.0.0"
 
+globby@11.0.1, globby@^11.0.1:
```
Comment (alexandcote) [inline yarn.lock:4430]: You can probably dedupe this too.

File: yarn.lock [hunk 6]
```diff
@@ -4945,6 +5196,13 @@ is-regex@^1.0.5:
   dependencies:
     has "^1.0.3"
 
+is-regex@^1.1.0:
```
Comment (alexandcote) [inline yarn.lock:5199]: dedupe

File: packages/eslint-plugin/lib/config/rules/jest.js (hunk 2 above)
Comment (BPScott) [reply packages/eslint-plugin/lib/config/rules/jest.js:49]: They seem reasonable, good spot

**Review by alexandcote — APPROVED** [review]

**Comment by BPScott** [comment]
https://github.com/Shopify/web/pull/31363 is a web PR that contains these changes.

444 new linting errors are introduced.

- 181 due to the introduction of `jest/no-deprecated-functions` and can be trivially fixed by replacing "require.THING" with "jest.THING".
- 133 due to `babel/no-unused-expressions` no liking the usage of optional chaining. This is somewhat annoying but updating ESLint to >= 7.5 (which adds optional chaining support) should fix them
- 52 due to `@typescript-eslint/naming-convention` changes already introduced in prior PRs 
- 27 due to us  removing `@typescript-eslint/camelcase` usage in favor of `@typescript-eslint/naming-convention`
- 29 due to `react/no-unused-prop-types`

A web build containing these changes is https://buildkite.com/shopify/web-ci-builder/builds/244536#b110f57c-5714-43d6-97c1-5c04a6b2de1f

**Pull request merged**